beacon-chain/rpc/beaconv1/blocks.go

```diff
@@ -24,20 +118,234 @@ func (bs *Server) ListBlockHeaders(ctx context.Context, req *ethpb.BlockHeadersR
+		return nil, status.Errorf(codes.Internal, "Could not process beacon block: %v", err)
+	}
+
+	return &ptypes.Empty{}, nil
 }
 
 // GetBlock retrieves block details for given block id.
 func (bs *Server) GetBlock(ctx context.Context, req *ethpb.BlockRequest) (*ethpb.BlockResponse, error) {
-	return nil, errors.New("unimplemented")
+	blk, err := bs.blockFromBlockID(ctx, req.BlockId)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get block from block ID: %v", err)
+	}
+	if blk == nil {
+		return nil, status.Errorf(codes.Internal, "Could not find requested block")
+	}
+
+	v1Block, err := v1alpha1ToV1Block(blk)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not convert block to v1")
+	}
+
+	return &ethpb.BlockResponse{
+		Data: &ethpb.BeaconBlockContainer{
+			Message:   v1Block.Block,
+			Signature: blk.Signature,
+		},
+	}, nil
 }
 
 // GetBlockRoot retrieves hashTreeRoot of BeaconBlock/BeaconBlockHeader.
 func (bs *Server) GetBlockRoot(ctx context.Context, req *ethpb.BlockRequest) (*ethpb.BlockRootResponse, error) {
-	return nil, errors.New("unimplemented")
+	switch string(req.BlockId) {
+	case "head":
+		root, err := bs.HeadFetcher.HeadRoot(ctx)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not retrieve head block: %v", err)
+		}
+		if root == nil {
+			return nil, status.Errorf(codes.Internal, "No head root was found")
+		}
+		return &ethpb.BlockRootResponse{
+			Data: &ethpb.BlockRootContainer{
+				Root: root,
+			},
+		}, nil
+	case "finalized":
+		finalized := bs.FinalizationFetcher.FinalizedCheckpt()
+		if finalized == nil {
+			return nil, status.Errorf(codes.Internal, "No finalized root was found")
+		}
+		return &ethpb.BlockRootResponse{
+			Data: &ethpb.BlockRootContainer{
+				Root: finalized.Root,
+			},
+		}, nil
+	case "genesis":
+		blk, err := bs.BeaconDB.GenesisBlock(ctx)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not retrieve blocks for genesis slot: %v", err)
+		}
+		if blk == nil {
+			return nil, status.Error(codes.Internal, "Could not find genesis block")
+		}
+		root, err := blk.Block.HashTreeRoot()
+		if err != nil {
+			return nil, status.Error(codes.Internal, "Could not hash genesis block")
+		}
+		return &ethpb.BlockRootResponse{
+			Data: &ethpb.BlockRootContainer{
+				Root: root[:],
+			},
+		}, nil
+	default:
+		if len(req.BlockId) == 32 {
+			return &ethpb.BlockRootResponse{
+				Data: &ethpb.BlockRootContainer{
+					Root: req.BlockId,
+				},
+			}, nil
+		} else {
+			slot := bytesutil.BytesToUint64BigEndian(req.BlockId)
+			roots, err := bs.BeaconDB.BlockRoots(ctx, filters.NewFilter().SetStartSlot(slot).SetEndSlot(slot))
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "Could not retrieve blocks for slot %d: %v", slot, err)
+			}
+
+			numRoots := len(roots)
+			if numRoots == 0 {
+				return nil, nil
+			}
+			root := roots[0]
```

Comment:
What about when there is more than 1 root? Shouldn't this return the "canonical" one only?

Reply:
How can I tell if its canonical?